File: DAO/DesignPattern/typescript_demo/factory.ts
// 计算机 - 简单工厂模式
// 运算类 - 基类
class Operation{
    private _numberA: number = 0;
    private _numberB: number = 0;

    set numberA(num:number){
        this._numberA = num;
    }

    get numberA(): number{
        return this._numberA;
    }

    set numberB(num:number){
        this._numberB = num;
    }

    get numberB(): number{
        return this._numberB;
    }

    getResult(): number{
        let result: number = 0;
        return result;
    }
}

// 具体运算类
class OperationAdd extends Operation{
    getResult(): number{
        let result: number = 0;
        result = this.numberA + this.numberB;
        return result;
    }
}

class OperationSub extends Operation{
    getResult(): number{
        let result: number = 0;
        result = this.numberA - this.numberB;
        return result;
    }
}

class OperationMul extends Operation{
    getResult(): number{
        let result: number = 0;
        result = this.numberA * this.numberB;
        return result;
    }
}

class OperationDiv extends Operation{
    getResult(): number{
        let result: number = 0;
        if (this.numberB == 0){
            throw new Error('除数不能为0。');
        }
        result = this.numberA / this.numberB;
        return result;
    }
}

// 运算工厂 - 进行具体实例化
class OperationFactory{
    static createOperate(operate: string): Operation{
        // let oper: Operation = null; // 这样不行, 必须要赋一个 Operation类型的值
        // let oper: Operation = new OperationAdd(); 
        // 只要保证 oper 绝对可以被赋值, 这里也可以不赋初值
        let oper: Operation;
        switch(operate){
            case '+':
                oper = new OperationAdd();
                break;
            case '-':
                oper = new OperationSub();
                break;
            case '*':
                oper = new OperationMul();
                break;
            case '/':
                oper = new OperationDiv();
                break;
            // 加个default, 保证了oper在switch语句内必然可以被赋值
            default:
                oper = new OperationAdd();
        }
        return oper;
    }
}

// 客户端代码
function factory_program(numa: number, operate: string, numb: number): number{
    let oper: Operation = OperationFactory.createOperate(operate);
    oper.numberA = numa;
    oper.numberB = numb;
    let result: number = oper.getResult();
    return result;
}

console.log(factory_program(1, '+', 2));
console.log(factory_program(4, '*', 2));
console.log(factory_program(1, '-', 2));
console.log(factory_program(6, '/', 2));

File: DAO/DesignPattern/origin/factory.cs
// 计算机 - 简单工厂模式
// 运算类
public class Operation
{
    private double _numberA = 0;
    private double _numberB = 0;
    

    // Getter && Setter
    // 数字A
    public double NumberA
    {
        get
        {
            return _numberA;
        }
        set
        {
            _numberA = value; // 这个value怎么来的
        }
    }

    // 数字B
    public double NumberB
    {
        get
        {
            return _numberB;
        }
        set
        {
            _numberB = value;
        }
    }

    // 得到运算结果
    public virtual double getResult()
    {
        double result = 0; 
        return result;
    }

    
}

// 加法类
class OperationAdd : Operation
{
    public override double getResult()
    {
        double result = 0; 
        result = NumberA + NumberB;
        return result;
    }
}

// 减法类
class OperationSub : Operation
{
    public override double getResult()
    {
        double result = 0;
        result = NumberA - NumberB;
        return result;
    }
}

// 乘法类
class OperationMul : Operation
{
    public override double getResult()
    {
        double result = 0;
        result = NumberA * NumberB;
        return result;
    }
}

// 除法类
class OperationDiv : Operation
{
    public override double getResult()
    {
        double result = 0;
        if (NumberB==0)
            throw new Exception("除数不能为0。");
        result = NumberA / NumberB;
        return result;
    }
}

// 运算类工厂 - 根据不同条件生成不同的类实例，每个实例都源自Operation类的子类, 继承了Operation类的属性
class OperationFactory
{
    public static Operation createOperate(string operate)
    {
        Operation oper = null;
        // 这就是多态的实现吧
        switch (operate)
        {
            case "+":
                {
                    oper = new OperationAdd();
                    break;
                }
            case "-":
                {
                    oper = new OperationSub();
                    break;
                }
            case "*":
                {
                    oper = new OperationMul();
                    break;
                }
            case "/":
                {
                    oper = new OperationDiv();
                    break;
                }
        }

        return oper;
    }
}

// 客户端代码
private void  Program(){
    Operation oper;
    oper = OperationFactory.createOperate("+");
    oper.NumberA = 1;
    oper.NumberB = 2;
    double result = oper.getResult();
}
